From 3d4cdfe16812cb302de07f08533079f97834bf18 Mon Sep 17 00:00:00 2001
From: Hui Tang <tanghui20@huawei.com>
Date: Fri, 27 May 2022 14:48:24 +0800
Subject: [PATCH 2189/2592] [Huawei] sched: Add feature 'UTIL_TASKGROUP' for
 dynamic affinity

Offering: HULK
hulk inclusion
category: bugfix
bugzilla: 186867

--------------------------------

If the feature enabled, util_avg of bottom-level tg is used,
otherwise, util_avg of rq's cfs_rq is used.

Signed-off-by: Hui Tang <tanghui20@huawei.com>
---
 kernel/sched/fair.c     | 17 ++++++++++++-----
 kernel/sched/features.h |  7 +++++++
 2 files changed, 19 insertions(+), 5 deletions(-)

diff --git a/kernel/sched/fair.c b/kernel/sched/fair.c
index fb3f9e004b59..bf5e842d79d5 100644
--- a/kernel/sched/fair.c
+++ b/kernel/sched/fair.c
@@ -6901,6 +6901,15 @@ static inline bool prefer_cpus_valid(struct task_struct *p)
 	       cpumask_subset(p->prefer_cpus, p->cpus_ptr);
 }
 
+static inline unsigned long taskgroup_cpu_util(struct task_group *tg,
+					       int cpu)
+{
+	if (tg->se[cpu] && sched_feat(DA_UTIL_TASKGROUP))
+		return tg->se[cpu]->avg.util_avg;
+
+	return cpu_util(cpu);
+}
+
 /*
  * set_task_select_cpus: select the cpu range for task
  * @p: the task whose available cpu range will to set
@@ -6931,13 +6940,11 @@ static void set_task_select_cpus(struct task_struct *p, int *idlest_cpu,
 	rcu_read_lock();
 	tg = task_group(p);
 	for_each_cpu(cpu, p->prefer_cpus) {
-		if (unlikely(!tg->se[cpu]))
-			continue;
-
 		if (idlest_cpu && available_idle_cpu(cpu)) {
 			*idlest_cpu = cpu;
 		} else if (idlest_cpu) {
-			spare = (long)(capacity_of(cpu) - tg->se[cpu]->avg.util_avg);
+			spare = (long)(capacity_of(cpu) -
+				taskgroup_cpu_util(tg, cpu));
 			if (spare > min_util) {
 				min_util = spare;
 				*idlest_cpu = cpu;
@@ -6952,7 +6959,7 @@ static void set_task_select_cpus(struct task_struct *p, int *idlest_cpu,
 			return;
 		}
 
-		util_avg_sum += tg->se[cpu]->avg.util_avg;
+		util_avg_sum += taskgroup_cpu_util(tg, cpu);
 		tg_capacity += capacity_of(cpu);
 	}
 	rcu_read_unlock();
diff --git a/kernel/sched/features.h b/kernel/sched/features.h
index 97ed11bd25e7..b358bbcb800b 100644
--- a/kernel/sched/features.h
+++ b/kernel/sched/features.h
@@ -100,3 +100,10 @@ SCHED_FEAT(UTIL_EST_FASTUP, true)
 
 SCHED_FEAT(ALT_PERIOD, true)
 SCHED_FEAT(BASE_SLICE, true)
+
+#ifdef CONFIG_QOS_SCHED_DYNAMIC_AFFINITY
+/*
+ * Use util_avg of bottom-Level taskgroup
+ */
+SCHED_FEAT(DA_UTIL_TASKGROUP, true)
+#endif
-- 
2.17.1

